compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3047,31 +3048,63 @@ TR::Register *OMR::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::C
       {
       TR::InstOpCode::Mnemonic op = TR::InstOpCode::bad;
 
-      if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicAddSymbol))
+      if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicAddSymbol))
          {
          op = node->getChild(1)->getDataType().isInt32() ? TR::InstOpCode::LADD4MemReg : TR::InstOpCode::LADD8MemReg;
          }
-      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAddSymbol))
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAddSymbol))
          {
-         op = node->getChild(1)->getDataType().isInt32() ? TR::InstOpCode::LXADD4MemReg : TR::InstOpCode::LXADD8MemReg;
+         auto dataType = node->getChild(1)->getDataType();
+         if (dataType.isInt8())
+            op = TR::InstOpCode::LXADD1MemReg;
+         else if (dataType.isInt16())
+            op = TR::InstOpCode::LXADD2MemReg;
+         else if (dataType.isInt32())
+            op = TR::InstOpCode::LXADD4MemReg;
+         else
+            op = TR::InstOpCode::LXADD8MemReg;
+         }
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd8BitSymbol))
+         {
+         op = TR::InstOpCode::LXADD1MemReg;
+         }
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd16BitSymbol))
+         {
+         op = TR::InstOpCode::LXADD2MemReg;
          }
-      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd32BitSymbol))
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd32BitSymbol))
          {
          op = TR::InstOpCode::LXADD4MemReg;
          }
-      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd64BitSymbol))
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicFetchAndAdd64BitSymbol))
          {
          op = TR::InstOpCode::LXADD8MemReg;
          }
-      else if (comp->getSymRefTab()->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwapSymbol))
+      else if (symRefTab->isNonHelper(SymRef, TR::SymbolReferenceTable::atomicSwapSymbol))
+         {
+         auto dataType = node->getChild(1)->getDataType();
+         if (dataType.isInt8())
+            op = TR::InstOpCode::XCHG1MemReg;
+         else if (dataType.isInt16())
+            op = TR::InstOpCode::XCHG2MemReg;
+         else if (dataType.isInt32())
+            op = TR::InstOpCode::XCHG4MemReg;
```

Comment:
As in the case of `atomicFetchAndAddSymbol` above, other platforms that currently support the `atomicSwapSymbol` non-helper aren't prepared to deal with 8 and 16 bit operands. I would suggest that the version of `atomicSwapSymbol` that doesn't specify a number of bits only be used for backwards compatibility for 32 and 64 bit operands for now, as that's all that is guaranteed across platforms. Then when that generic symbol is no longer in use by downstream projects, it can be removed from OMR completely.

Reply:
I think that's a good idea. I'll leave a comment expressing this as well in the code